### PR TITLE
Add chmod handling to COPY/ADD

### PR DESCRIPTION
**Pull request by jmguzik**
In order to properly implement chmod in buildah this addition is needed.

Part of WIP:
https://github.com/containers/buildah/pull/3004

### CODE REVIEW

**Comment by jmguzik** [comment]
@rhatdan FYI

**Comment by jmguzik** [comment]
@rhatdan @nalind @smarterclayton could you mark this change as ok to test?

**Comment by jmguzik** [comment]
@nalind since all jobs succeded, and you responded do fast, could I ask you to review this one?

File: dispatchers_test.go [hunk 1]
```diff
@@ -2,11 +2,12 @@ package imagebuilder
 
 import (
 	"errors"
-	"github.com/containerd/containerd/platforms"
 	"reflect"
 	"sort"
 	"testing"
 
+	"github.com/containerd/containerd/platforms"
+
```
Comment (rhatdan) [inline dispatchers_test.go:10]: Nit: remove blank line.

**Review by rhatdan — APPROVED** [review]
Small nit, but LGTM

File: dispatchers_test.go (hunk 1 above)
Comment (jmguzik) [reply dispatchers_test.go]: Line removed, thanks.

**Comment by nalind** [comment]
Tests that use `--chmod=888` look kind of iffy, in that they seem to only be testing that 888 doesn't get interpreted as 777 or 755, otherwise it looks correct.

**Comment by jmguzik** [comment]
@nalind What would be your suggestion? Change them (could you propose something?) or get rid of them?

**Comment by rhatdan** [comment]
@jmguzik Doesn't the `888` value or `bogus` trigger a real error?

**Comment by jmguzik** [comment]
> @jmguzik Doesn't the `888` value or `bogus` trigger a real error?

No, right now it acts exactly like chown implementation. Does not check anything, just forwards.
In fact 888 could be replaced by any other string and if checked in params, and it would be validated as correct one. I just followed chown implementation of bad values and chose 888 to reflect it in test. Could be any other string.
In my understanding it is not the place to check, however if you would like to see this check here please let me know. It would land probably in add and copy functions in builder.

I think what @nalind meant is just to remove 888 to replace it with any other value from range to remve "iffy look". From the testing perspective, nothing will change. The same applies to chown impl.

**Comment by rhatdan** [comment]
Ok do that and I will merge.

**Comment by jmguzik** [comment]
> Ok do that and I will merge.

Done

File: dispatchers.go [hunk 2]
```diff
@@ -179,13 +187,19 @@ func dispatchCopy(b *Builder, args []string, attributes map[string]bool, flagArg
 		switch {
 		case strings.HasPrefix(arg, "--chown="):
 			chown = strings.TrimPrefix(arg, "--chown=")
+		case strings.HasPrefix(arg, "--chmod="):
+			chmod = strings.TrimPrefix(arg, "--chmod=")
+			_, err = strconv.ParseUint(chmod, 8, 32)
+			if err != nil {
+				return errChmodConversion(chmod)
+			}
```
Comment (TomSweeneyRedHat) [inline dispatchers.go:195]: nit, you might consider renaming errChmodConversion to something like checkChmodConversion then putting the ParseUnit() call into and then return err or nil

**Review by TomSweeneyRedHat — COMMENTED** [review]
One nit/suggestion.
Otherwise LGTM, but would like a head nod from @nalind

File: dispatchers.go (hunk 2 above)
Comment (jmguzik) [reply dispatchers.go:195]: This is exactly what I had (even the same name), but then spotted err bla bla... funtions at the bottom and decided to follow this convention. Anyway, I will fix it right away.

**Comment by jmguzik** [comment]
Done. Added one more check in tests to prevent a problem with null.
